core/translate/optimizer.rs

```diff
@@ -500,6 +501,46 @@ fn push_scan_direction(operator: &mut SourceOperator, direction: &Direction) {
     }
 }
 
+fn eliminate_between(
+    operator: &mut SourceOperator,
+    where_clauses: &mut Option<Vec<ast::Expr>>,
+) -> Result<()> {
+    if let Some(predicates) = where_clauses {
+        *predicates = predicates.drain(..).map(convert_between_expr).collect();
```

Comment:
could we mutate in place as much as possible? i don't know if rust compiler is somehow able to reuse the allocation but i don't think there's a reason to `drain().map().collect()` when we can mutably iterate and modify in place.

same with the `convert_between_expr` function

otherwise I think this is a good solution and as mentioned, we can take care of the extra bytecode instruction later

Reply:
> i don't know if rust compiler is somehow able to reuse the allocation

This is the benefit of drain. We don't need to worry about lifecycle issues. If we use reference, when we want to create a new Expr, we need to clone the field of the original Expr.

Drain does not actually consume `Vec`, in other words it reuses the allocated.
[The Rustonomicon Drain](https://doc.rust-lang.org/nomicon/vec/vec-drain.html)
[Stackoverflow-when should i use drain vs into iter](https://stackoverflow.com/a/27882895)

Reply:
that's cool 👍 one last thing, can you write a couple of basic tcl tests in where.test or a similar location?

Reply:
If we let `convert_between_expr` accept a mutable reference, consider a scenario like this:
```
fn convert_between_expr(expr: &mut ast::Expr) {
    match expr {
         ast::Expr::Between {
            lhs,
            not,
            start,
            end,
        } => {
            // Oh no! we can not move out of *start/*lhs/*end which is behind a mutable reference
            let lower_bound = ast::Expr::Binary(*start, ast::Operator::LessEquals, lhs.clone());
            let upper_bound = ast::Expr::Binary(*lhs, ast::Operator::LessEquals, *end);
            // We will eventually clone them one by one here
    }
}
```

Reply:
I think in those cases you could use `std::mem::take` or `std::mem::swap` instead of cloning

Reply:
But `Expr` does not implement `Default` trait.
[Expr enum](https://docs.rs/sqlite3-parser/latest/src/sqlite3_parser/parser/ast/mod.rs.html#272-392)
[std::mem::take](https://doc.rust-lang.org/std/mem/fn.take.html)

Reply:
Yeah actually I remember now that I've been using this hack to do effectively the same thing

```
trait TakeOwnership {
    fn take_ownership(&mut self) -> Self;
}

impl TakeOwnership for ast::Expr {
    fn take_ownership(&mut self) -> Self {
        std::mem::replace(self, ast::Expr::Literal(ast::Literal::Null))
    }
}
```

anyway my main concern was about the heap allocation but since that doesn't happen I don't mind the current impl. Could you add a few tests as I asked above?

Reply:
> that's cool 👍 one last thing, can you write a couple of basic tcl tests in where.test or a similar location?

No problem.

Reply:
Added some very basic tests